[Notifier] Fix thread key in GoogleChat bridge

romain-jacquart opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #57323
| License       | MIT

Hello,

Google Chat API [has deprecated](https://developers.google.com/workspace/chat/api/reference/rest/v1/spaces.messages/create#query-parameters) the use of `threadKey` query parameter in favor of `thread.threadKey` in request's body.

In order for the thread key value to not be ignored, a [new query parameter `messageReplyOption`](https://developers.google.com/workspace/chat/api/reference/rest/v1/spaces.messages/create#messagereplyoption) has been added which controls the way the key is handled.
I chose **not to expose** the values used by this new parameter as I wanted to focus this PR on the actual fix.
To do so, I've forced the value of `messageReplyOption` to `REPLY_MESSAGE_FALLBACK_TO_NEW_THREAD` which replicates the original behavior of this bridge.

OskarStark commented:
Please target 7.2 branch for your PR as this is a new feature. thanks

OskarStark commented:
It introduces new code to handle the new threadKey stuff. So this should go into 7.2 with he deprecation and the deprecation layer be removed in 8.0.

Simone who needs this should upgrade to 7.2 then

fabpot commented:
I agree that this should go in 7.2.

GromNaN commented on src/Symfony/Component/Notifier/Bridge/GoogleChat/README.md:
```diff
@@ -14,7 +14,7 @@ where:
  - `ACCESS_TOKEN` is your Google Chat access token
  - `SPACE` is the Google Chat space
- - `THREAD_KEY` is the Google Chat message thread to group messages into a single thread (optional)
+ - ~~`THREAD_KEY` is the Google Chat message thread to group messages into a single thread (optional)~~ (deprecated)
```
Why deprecating this feature? I think it's still a good option to be able to choose a default thread for all notifications.

romain-jacquart replied:
I assumed this DSN should closely match the actual API's endpoint. This has been reverted back.

GromNaN commented on src/Symfony/Component/Notifier/Bridge/GoogleChat/GoogleChatOptions.php:
```diff
@@ -82,18 +81,41 @@ public function text(string $text): static
      * @return $this
      */
-    public function setThreadKey(?string $threadKey): static
+    public function threadKey(string $threadKey): static
```
`$threadKey` should be nullable to unset the value.

```suggestion
    public function threadKey(?string $threadKey): static
```

romain-jacquart replied:
This method no long exists.

GromNaN commented on src/Symfony/Component/Notifier/Bridge/GoogleChat/GoogleChatOptions.php:
```diff
@@ -82,18 +81,41 @@ public function text(string $text): static
+    /**
+     * @return $this
+     *
+     * @deprecated since Symfony 7.2, to be removed in 8.0
```
Why do you need to deprecate this method? The new function `threadKey` serves the same purpose.

romain-jacquart replied:
This change was made to adopt the same "naming convention" of the other methods in this class.

The new `threadKey()` method served the same purpose because it was a simple rename of the setter and I wanted to keep the backward compatibility here.

This has been reverted back and the setter has been kept.

GromNaN reviewed (changes requested):
Thanks for catching this deprecation in the Google API.

> This PR keeps the backward compatibility regarding the threadKey declaration (in DSN) and the GoogleChatOptions::setThreadKey() method.

I think this PR does too many things. We could just modify the HTTP request, and leave the options and methods untouched.

romain-jacquart commented:
> > This PR keeps the backward compatibility regarding the threadKey declaration (in DSN) and the GoogleChatOptions::setThreadKey() method.
> 
> I think this PR does too many things. We could just modify the HTTP request, and leave the options and methods untouched.

Thanks for your time reviewing this change.

I thought this was a good opportunity to update the options, but I guess this made too much noise around the actual fix.

GromNaN commented on src/Symfony/Component/Notifier/Bridge/GoogleChat/GoogleChatTransport.php:
```diff
@@ -52,10 +52,9 @@ public function __construct(
     public function __toString(): string
     {
-        return sprintf('googlechat://%s/%s%s',
+        return sprintf('googlechat://%s/%s',
```
This needs to be reverted too. It's the Symfony DSN to configure this bridge, not the URL to the Google API Endpoint.

romain-jacquart replied:
Whoopsie, good catch, thanks! 👍

GromNaN commented:
Thanks for simplifying this PR. I think we can now consider it as a bugfix to guarantee compatibility with the Google API in the future and merge in 5.4 @symfony/mergers?

fabpot reviewed (approved):
Works for me as a bug fix for 5.4.
@romain-jacquart Can you remove the CHANGELOG change and rebase on 5.4?

GromNaN reviewed (approved):
LGTM once rebased on 5.4

romain-jacquart committed:
[Notifier] Fix thread key in GoogleChat bridge

Google Chat API has deprecated the use of `threadKey` query parameter
in favor of `thread.threadKey` in request's body.

romain-jacquart commented:
> Works for me as a bug fix for 5.4. @romain-jacquart Can you remove the CHANGELOG change and rebase on 5.4?

The branch has been rebased on 5.4 and CHANGELOG changes have been removed 👍.

The rebase has introduced a slight change regarding a variable's name (`$options` vs `$opts`). Will it be a problem?

Thanks!

fabpot commented:
Thank you @romain-jacquart.
